packages/agents-core/src/runImplementation.ts

```diff
@@ -336,6 +407,40 @@ export async function executeToolsAndSideEffects<TContext>(
   newItems = newItems.concat(functionResults.map((r) => r.runItem));
   newItems = newItems.concat(computerResults);
 
+  // run hosted MCP approval requests
+  if (processedResponse.mcpApprovalRequests.length > 0) {
+    for (const approvalRequest of processedResponse.mcpApprovalRequests) {
+      const toolData = approvalRequest.mcpTool
+        .providerData as ProviderData.HostedMCPTool<TContext>;
+      if (!toolData.onApproval) {
+        throw new UserError(
+          `Hosted remote MCP server tool (${toolData.serverLabel}) does not have an onApproval function`,
+        );
+      }
+      const approvalResult = await toolData.onApproval(
```

Comment:
This is the same design with Python SDK. Developers can pass onApproval callback when passing HostedMCPTool to Agent.